### PR TITLE
Graph components all to be done in React

### PR DESCRIPTION
**Pull request by ryanfan**
Currently only the `nodes` are done in React.  I wanted to make sure I have a good way to do this before continuing to work on `regions`, `bools`, `edges`, `region-labels`.

### CODE REVIEW

File: public/js/graph/postsetup.js.jsx [hunk 1]
```diff
@@ -0,0 +1,26 @@
+/*this was moved outside of setup.js*/
```
Comment (Ian-Stewart-Binks) [inline public/js/graph/postsetup.js.jsx]: I'm not certain that this comment adds anything useful to the file for future maintainers.

It is useful for reviewers of the PR, but I think that something like that should exist in either a description of what the PR does, or a line note.

By the way, why was it moved?

File: public/js/graph/parse_graph.js [hunk 2]
```diff
@@ -14,8 +14,9 @@ function buildGraph(gId) {
     nodes = [];
 
     $('.node').each(function () {
-        makeNode('AND', $(this).attr('id'));
+        makeNode('AND', $(this).attr('id'));
```
Comment (Ian-Stewart-Binks) [inline public/js/graph/parse_graph.js]: Bad whitespace.

File: hs/Scripts.hs [hunk 3]
```diff
@@ -71,7 +71,9 @@ graphScripts = do
          "/static/js/graph/sidebar/focus_descriptions.js",
          "/static/js/common/export/export.js"])
     H.script ! A.type_ "text/jsx" ! A.src "/static/js/common/react_modal.js.jsx" $ ""
-
+    H.script ! A.type_ "text/jsx" ! A.src "/static/js/graph/react_node.js.jsx" $ ""
+    H.script ! A.type_ "text/jsx" ! A.src "/static/js/graph/postsetup.js.jsx" $ ""
+
```
Comment (Ian-Stewart-Binks) [inline hs/Scripts.hs]: Unnecessary whitespace.

File: public/js/graph/react_node.js.jsx [hunk 4]
```diff
@@ -0,0 +1,113 @@
+function renderReactGraph() {
+    React.render(
+        <ReactSVG width={1195} height={650}/>,
+        document.getElementById('react-graph')
+    );
+}
+
+function getAttributes(node_named_map) {
```
Comment (Ian-Stewart-Binks) [inline public/js/graph/react_node.js.jsx]: Why are these variables using the underscore naming convention?

File: public/js/graph/react_node.js.jsx [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    });
+    return attrs;
+}
+
+function getStyles(styles_strings) {
+    if (!styles_strings){
+        return {};
+    }
+    var styles = {};
+        styles_strings.split(";").map(function(key, value){
```
Comment (Ian-Stewart-Binks) [inline public/js/graph/react_node.js.jsx]: The indentation here appears to be off, and there's a space missing between here.

File: public/js/graph/postsetup.js.jsx (hunk 1 above)
Comment (ryanfan) [inline public/js/graph/postsetup.js.jsx]: jsx files are transformed in-browser, so if a js file need a function from a jsx file that wasn't transformed yet, it wouldn't find it.  By moving `$(document).read` into a jsx file, it would mean all the jsx files would be transformed before any code will be run.

File: public/js/graph/react_node.js.jsx (hunk 5 above)
Comment (ryanfan) [inline public/js/graph/react_node.js.jsx]: Will fix the styles in the next commit.  Thanks again.

File: public/js/graph/react_node.js.jsx [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+                var child_entry = {};
+                child_entry["attributes"] = getAttributes(child.attributes);
+                child_entry["style"] = getStyles(child_entry["attributes"]["style"]);
+                //innerHTML is just for the text tag
+                child_entry["innerHTML"] = child.innerHTML;
+                entry["children"].push(child_entry);
+            });
+            dict_list.push(entry);
+        });
+        this.setState({nodes_list:arr});
```
Comment (david-yz-liu) [inline public/js/graph/react_node.js.jsx]: Remove this line.

File: public/js/graph/react_node.js.jsx [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+            nodes_list: []
+        };
+    },
+    
+    componentDidMount: function(){
+        var dict_list = [];
+        
+        $('.node').map(function(key, value) {
+            var entry = {};
+            entry["attributes"] = getAttributes(value.attributes);
```
Comment (david-yz-liu) [inline public/js/graph/react_node.js.jsx]: Use single quotes for strings.

File: public/js/graph/react_node.js.jsx [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    React.render(
+        <ReactSVG width={1195} height={650}/>,
+        document.getElementById('react-graph')
+    );
+}
+
+function getAttributes(node_named_map) {
+    var attrs = [];
+    //Traversing a NodeNamedMap type
+    Array.prototype.slice.call(node_named_map).forEach(function(item) {
```
Comment (david-yz-liu) [inline public/js/graph/react_node.js.jsx]: Alright, try doing 

```
for (var i = 0; i < node_named_map.length; i++) {
  var item = node_named_map[i];
  attrs[item.name] = item.value;
```

You should be able to loop through NamedNodeMaps like an array.

File: public/js/graph/react_node.js.jsx [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        attrs[item.name] = item.value;
+    });
+    return attrs;
+}
+
+function getStyles(styles_strings) {
+    if (!styles_strings){
```
Comment (david-yz-liu) [inline public/js/graph/react_node.js.jsx]: I suspect this block is redundant.

File: public/js/graph/react_node.js.jsx [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        var dict_list = [];
+        
+        $('.node').map(function(key, value) {
+            var entry = {};
+            entry["attributes"] = getAttributes(value.attributes);
+            //assume no styles here
+            entry["style"] = getStyles(entry["attributes"]["style"]);
+            entry["children"] = [];
+            //value.children is an HTML collection, converting to array here
+            var children_arr = [].slice.call(value.children);
```
Comment (david-yz-liu) [inline public/js/graph/react_node.js.jsx]: I think you can just do `Array(value.children)` here.

File: public/js/graph/react_node.js.jsx [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    getInitialState: function(){
+        return {
+            nodes_list: []
+        };
+    },
+    
+    componentDidMount: function(){
+        var dict_list = [];
+        
+        $('.node').map(function(key, value) {
```
Comment (david-yz-liu) [inline public/js/graph/react_node.js.jsx]: use `element` instead of `value`

File: public/js/graph/react_node.js.jsx [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+            var entry = {};
+            entry["attributes"] = getAttributes(value.attributes);
+            //assume no styles here
+            entry["style"] = getStyles(entry["attributes"]["style"]);
+            entry["children"] = [];
+            //value.children is an HTML collection, converting to array here
+            var children_arr = [].slice.call(value.children);
+            //assume only one level of children
+            //for each children
+            children_arr.map(function(child) {
```
Comment (david-yz-liu) [inline public/js/graph/react_node.js.jsx]: Use `forEach` instead of `map`

File: public/js/graph/react_node.js.jsx [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+                </rect>
+                {//this.props.node.children is an HTMLCollection, not an array
+                this.props.children.slice(1).map(function(text_tag) {
+                    return <text {... text_tag["attributes"]}>{text_tag["innerHTML"]}</text>;
+                })
+                }
+            </g>
+        );
+    }
+});
```
Comment (david-yz-liu) [inline public/js/graph/react_node.js.jsx]: Add a newline here

File: public/js/graph/postsetup.js.jsx (hunk 1 above)
Comment (david-yz-liu) [inline public/js/graph/postsetup.js.jsx]: @Ian-Stewart-Binks is right, we should remove this comment.

File: hs/Scripts.hs [hunk 14]
```diff
@@ -71,7 +71,9 @@ graphScripts = do
          "/static/js/graph/sidebar/focus_descriptions.js",
          "/static/js/common/export/export.js"])
     H.script ! A.type_ "text/jsx" ! A.src "/static/js/common/react_modal.js.jsx" $ ""
-
+    H.script ! A.type_ "text/jsx" ! A.src "/static/js/graph/react_node.js.jsx" $ ""
```
Comment (Ian-Stewart-Binks) [inline hs/Scripts.hs]: The `toScript` function appears to be gaining ambiguity here, and this section of code is not consistent. Should we remove the `toScript` function, or create a sister `test/jsx` `toScript` function as well?

File: public/js/graph/react_node.js.jsx [hunk 15]
```diff
@@ -0,0 +1,109 @@
+function renderReactGraph() {
+    React.render(
+        <ReactSVG width={1195} height={650}/>,
+        document.getElementById('react-graph')
+    );
+}
+
+function getAttributes(svgAttributes) {
+    var attrs = [];
```
Comment (Ian-Stewart-Binks) [inline public/js/graph/react_node.js.jsx]: Missing use strict.

File: public/js/graph/react_node.js.jsx [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+    },
+    
+    componentDidMount: function(){
+        var dict_list = [];
+        
+        $('.node').map(function(key, value) {
+            var entry = {};
+            entry["id"] = value.id;
+            entry["attributes"] = getAttributes(value.attributes);
+            //assume no styles here
```
Comment (Ian-Stewart-Binks) [inline public/js/graph/react_node.js.jsx]: Is there any trade-off to assuming no styles here? Why is this being done? Is there any way to add in more helpful information to this comment?

File: public/js/graph/react_node.js.jsx [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+            //value.children is an HTML collection, converting to array here
+            var children_arr = [].slice.call(value.children);
+            //assume only one level of children
+            //for each children
+            children_arr.map(function(child) {
+                var child_entry = {};
+                child_entry["attributes"] = getAttributes(child.attributes);
+                child_entry["style"] = getStyles(child_entry["attributes"]["style"]);
+                //innerHTML is just for the text tag
+                child_entry["innerHTML"] = child.innerHTML;
```
Comment (Ian-Stewart-Binks) [inline public/js/graph/react_node.js.jsx]: This is just storing the `child.innerHTML` value in the `child_entry`, correct? Would it be more useful to use a more descriptive name, rather than a comment, here?

File: public/js/graph/react_node.js.jsx [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+function getAttributes(svgAttributes) {
+    var attrs = [];
+    //Traversing a NodeNamedMap type
+    Array.prototype.slice.call(svgAttributes).forEach(function(item) {
+        attrs[item.name] = item.value;
+    });
+    return attrs;
+}
+
+function getStyles(styles_strings) {
```
Comment (Ian-Stewart-Binks) [inline public/js/graph/react_node.js.jsx]: camelCase.

File: public/js/graph/react_node.js.jsx [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+    var styles = {};
+    if (!styles_strings) {
+        return styles;
+    }
+    styles_strings.split(";").map(function(key, value){
```
Comment (Ian-Stewart-Binks) [inline public/js/graph/react_node.js.jsx]: Space.

File: public/js/graph/react_node.js.jsx [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+        if (key) {
+            styles[key.substring(0, key.indexOf(':'))] = key.substring(key.indexOf(':')+1);
```
Comment (Ian-Stewart-Binks) [inline public/js/graph/react_node.js.jsx]: Spaces.

File: public/js/graph/react_node.js.jsx [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+                {/*<ReactBools/>*/}
+                {/*<ReactEdges/>*/}
+                {/*<ReactRegionLabels/>*/}
+            </svg>
+        );
+    }
+});
+
+var ReactNodes = React.createClass({
+    getInitialState: function(){
```
Comment (Ian-Stewart-Binks) [inline public/js/graph/react_node.js.jsx]: Space.

File: public/js/graph/react_node.js.jsx [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+        return {
+            nodes_list: []
+        };
+    },
+    
+    componentDidMount: function(){
```
Comment (Ian-Stewart-Binks) [inline public/js/graph/react_node.js.jsx]: Space.

File: public/js/graph/react_node.js.jsx [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+                    return <ReactNode key={entry["id"]} attributes={entry["attributes"]} style={entry["style"]} children={entry["children"]}/>
+                })}
+            </g>
+        );
+    }
+});
+
+//Kept as separate component in case it may be needed later
+var ReactNode = React.createClass({
+    render: function(){
```
Comment (Ian-Stewart-Binks) [inline public/js/graph/react_node.js.jsx]: Space.

File: public/js/graph/react_node.js.jsx (hunk 10 above)
Comment (ryanfan) [inline public/js/graph/react_node.js.jsx]: I learned to do this from here:
http://stackoverflow.com/questions/222841/most-efficient-way-to-convert-an-htmlcollection-to-an-array
Doing `Array(value.children)` would give me an `[HTMLCollection[2]]`, but I wanted [rect, text] instead.

File: hs/Scripts.hs (hunk 14 above)
Comment (ryanfan) [inline hs/Scripts.hs]: I'm not exactly sure what do do here.
@david-yz-liu Can you give some advice please?

File: public/js/graph/react_node.js.jsx (hunk 17 above)
Comment (ryanfan) [inline public/js/graph/react_node.js.jsx]: I'm not sure what you want here.  innerHTML is just the text within the `<text></text>`, `<rect></rect>` does not have any text inside.  It was also assumed that there was only one level of children according to the structure of the SVG, so there would be no more tags within these tags.

File: hs/Scripts.hs (hunk 14 above)
Comment (david-yz-liu) [inline hs/Scripts.hs]: @ryanfan You can leave this for now.
@Ian-Stewart-Binks You're right, creating a new `jsxScript` function makes sense, I think.

File: public/js/graph/react_node.js.jsx [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+//Kept as separate component in case it may be needed later
+var ReactNode = React.createClass({
+    render: function() {
+        //hard-coded className
+        return (
+            <g className={this.props.className} {... this.props.attributes} style={this.props.styles}>
+                <rect {... this.props.children[0]['attributes']} style={this.props.children[0]['style']}>
+                </rect>
+                {//this.props.node.children is an HTMLCollection, not an array
+                this.props.children.slice(1).map(function(text_tag, value) {
```
Comment (Ian-Stewart-Binks) [inline public/js/graph/react_node.js.jsx]: camelCase

File: public/js/graph/postsetup.js.jsx (hunk 1 above)
Comment (Ian-Stewart-Binks) [inline public/js/graph/postsetup.js.jsx]: @ryanfan What's the difference between the `postsetup` and `setup` files?

File: public/js/graph/react_node.js.jsx (hunk 17 above)
Comment (Ian-Stewart-Binks) [inline public/js/graph/react_node.js.jsx]: Ah, I just mean the `child_entry["innerHTML"]`. I may be mistaken, but the name here looks arbitrary, and it's main use is data storage to insert the text later on (please correct me if I'm wrong, I very well may be).

If that is the case, a more useful name may be better. The comment also seems to be fairly unnecessary, and a little ambiguous.

**Comment by ryanfan** [comment]
Still working on it, will remove the debugging `console.log` in later commit

**Comment by david-yz-liu** [comment]
@ryanfan I decided to merge this since it had gotten quite large; please make the minor fixes we discussed yesterday in a separate PR.
